x-pack/plugins/spaces/server/lib/spaces_client/spaces_client.ts

```diff
@@ -11,17 +11,23 @@ import { isReservedSpace } from '../../../common/is_reserved_space';
 import { Space } from '../../../common/model/space';
 import { SpacesAuditLogger } from '../audit_logger';
 import { ConfigType } from '../../config';
-import { GetSpacePurpose } from '../../../common/model/types';
+import { GetAllSpacesPurpose, GetSpaceResult } from '../../../common/model/types';
 
-const SUPPORTED_GET_SPACE_PURPOSES: GetSpacePurpose[] = [
+interface GetAllSpacesOptions {
+  purpose?: GetAllSpacesPurpose;
+  allowPartialAuthorization?: boolean;
```

Comment:
I know you tried to clear this with me beforehand, so I'm sorry about the following suggestion. I didn't have the idea until reviewing the implementation here.

Rather than allowing for partial authorization, what do you think about [optionally] augmenting the spaces with flags indicating which purposes the user is authorized for? I feel like that would be easier to understand as a consumer, and is a bit more flexible for the next time we need to add a purpose.

Here's a quick example I threw together. It likely needs some refinement, but it works in the handful of scenarios I tested:

```diff
diff --git a/x-pack/plugins/spaces/server/lib/spaces_client/spaces_client.ts b/x-pack/plugins/spaces/server/lib/spaces_client/spaces_client.ts
index 63979c47ca9..4241ff55633 100644
--- a/x-pack/plugins/spaces/server/lib/spaces_client/spaces_client.ts
+++ b/x-pack/plugins/spaces/server/lib/spaces_client/spaces_client.ts
@@ -15,7 +15,7 @@ import { GetAllSpacesPurpose, GetSpaceResult } from '../../../common/model/types
 
 interface GetAllSpacesOptions {
   purpose?: GetAllSpacesPurpose;
-  allowPartialAuthorization?: boolean;
+  includeAuthorizedPurposes?: boolean;
 }
 
 const SUPPORTED_GET_SPACE_PURPOSES: GetAllSpacesPurpose[] = [
@@ -35,7 +35,7 @@ const PURPOSE_PRIVILEGE_MAP: Record<
     authorization.actions.ui.get('savedObjectsManagement', 'copyIntoSpace'),
   ],
   findSavedObjects: (authorization) => {
-    return [authorization.actions.savedObject.get('config', 'find')];
+    return [authorization.actions.login, authorization.actions.savedObject.get('config', 'find')];
   },
   shareSavedObjectsIntoSpace: (authorization) => [
     authorization.actions.ui.get('savedObjectsManagement', 'shareIntoSpace'),
@@ -54,11 +54,15 @@ export class SpacesClient {
   ) {}
 
   public async getAll(options: GetAllSpacesOptions = {}): Promise<GetSpaceResult[]> {
-    const { purpose = DEFAULT_PURPOSE, allowPartialAuthorization = false } = options;
+    const { purpose = DEFAULT_PURPOSE, includeAuthorizedPurposes = false } = options;
     if (!SUPPORTED_GET_SPACE_PURPOSES.includes(purpose)) {
       throw Boom.badRequest(`unsupported space purpose: ${purpose}`);
     }
 
+    if (purpose !== DEFAULT_PURPOSE && includeAuthorizedPurposes) {
+      throw Boom.badRequest(`'purpose' cannot be supplied with 'includeAuthorizedPurposes'`);
+    }
+
     if (this.useRbac()) {
       // eslint-disable-next-line @typescript-eslint/naming-convention
       const { saved_objects } = await this.internalSavedObjectRepository.find({
@@ -70,26 +74,82 @@ export class SpacesClient {
 
       this.debugLogger(`SpacesClient.getAll(), using RBAC. Found ${saved_objects.length} spaces`);
 
-      let spaces: GetSpaceResult[] = saved_objects.map(this.transformSavedObjectToSpace);
-      const authorizedSpacesResult = await this.getAuthorizedSpaces(spaces, purpose);
-      const { username } = authorizedSpacesResult;
-      let required = authorizedSpacesResult.authorized;
-
-      if (allowPartialAuthorization && purpose !== DEFAULT_PURPOSE) {
-        // the default purpose 'any' has the least restrictive authorization checks;
-        // if we allow partial authorization, make an additional less-restrictive authorization check and modify the results accordingly
-        const { authorized } = await this.getAuthorizedSpaces(spaces, DEFAULT_PURPOSE);
-        const optional = required;
-        required = authorized;
-        spaces = spaces.map((space) => {
-          if (!optional.includes(space.id)) {
-            return { ...space, isPartiallyAuthorized: true };
+      const spaces: GetSpaceResult[] = saved_objects.map(this.transformSavedObjectToSpace);
+      const spaceIds = spaces.map((space: Space) => space.id);
+
+      const checkPrivileges = this.authorization!.checkPrivilegesWithRequest(this.request);
+
+      // Collect all privileges which need to be checked
+      const allPrivileges = Object.entries(PURPOSE_PRIVILEGE_MAP).reduce(
+        (acc, [getSpacesPurpose, privilegeFactory]) => {
+          return {
+            ...acc,
+            [getSpacesPurpose]: privilegeFactory(this.authorization!),
+          };
+        },
+        {} as Record<GetAllSpacesPurpose, string[]>
+      );
+
+      // Check all privileges against all spaces
+      const { username, privileges } = await checkPrivileges.atSpaces(spaceIds, {
+        kibana: Object.values(allPrivileges).flat(),
+      });
+
+      // Determine which purposes the user is authorized for within each space.
+      // Remove any spaces for which user is fully unauthorized.
+
+      const authorized = spaces
+        .map((space: Space) => {
+          if (!includeAuthorizedPurposes) {
+            // no need to go through all that work. Just check if they're authorized for the requested purpose
+            const requiredActions = PURPOSE_PRIVILEGE_MAP[purpose](this.authorization!);
+            const hasAllRequired = requiredActions.every((action) =>
+              privileges.kibana.some(
+                (privilege) =>
+                  privilege.resource === space.id &&
+                  privilege.privilege === action &&
+                  privilege.authorized === true
+              )
+            );
+
+            return hasAllRequired ? space : null;
+          }
+
+          let hasAnyAuthorization = false;
+          const authorizedPurposes = Object.entries(PURPOSE_PRIVILEGE_MAP).reduce(
+            (acc, [purposeKey, privilegeFactory]) => {
+              const requiredActions = privilegeFactory(this.authorization!);
+              const hasAllRequired = requiredActions.every((action) =>
+                privileges.kibana.some(
+                  (privilege) =>
+                    privilege.resource === space.id &&
+                    privilege.privilege === action &&
+                    privilege.authorized === true
+                )
+              );
+
+              hasAnyAuthorization = hasAnyAuthorization || hasAllRequired;
+
+              return {
+                ...acc,
+                [purposeKey]: hasAllRequired,
+              };
+            },
+            {} as Record<GetAllSpacesPurpose, boolean>
+          );
+
+          if (!hasAnyAuthorization) {
+            return null;
           }
-          return space;
-        });
-      }
 
-      if (required.length === 0) {
+          return {
+            ...space,
+            authorizedPurposes,
+          };
+        })
+        .filter((result: unknown) => result != null) as GetSpaceResult[];
+
+      if (authorized.length === 0) {
         this.debugLogger(
           `SpacesClient.getAll(), using RBAC. returning 403/Forbidden. Not authorized for any spaces for ${purpose} purpose.`
         );
@@ -97,14 +157,14 @@ export class SpacesClient {
         throw Boom.forbidden(); // Note: there is a catch for this in `SpacesSavedObjectsClient.find`; if we get rid of this error, remove that too
       }
 
-      this.auditLogger.spacesAuthorizationSuccess(username, 'getAll', required);
-      const filteredSpaces = spaces.filter((space: GetSpaceResult) => required.includes(space.id));
+      const authorizedSpaceIds = authorized.map((s) => s.id);
+
+      this.auditLogger.spacesAuthorizationSuccess(username, 'getAll', authorizedSpaceIds);
+
       this.debugLogger(
-        `SpacesClient.getAll(), using RBAC. returning spaces: ${filteredSpaces
-          .map((s) => s.id)
-          .join(',')}`
+        `SpacesClient.getAll(), using RBAC. returning spaces: ${authorizedSpaceIds.join(',')}`
       );
-      return filteredSpaces;
+      return authorized;
     } else {
       this.debugLogger(`SpacesClient.getAll(), NOT USING RBAC. querying all spaces`);
 
@@ -224,28 +284,6 @@ export class SpacesClient {
     return this.authorization != null && this.authorization.mode.useRbacForRequest(this.request);
   }
 
-  private async getAuthorizedSpaces(spaces: Space[], purpose: GetAllSpacesPurpose) {
-    const privilegeFactory = PURPOSE_PRIVILEGE_MAP[purpose];
-    const spaceIds = spaces.map((space: Space) => space.id);
-    const checkPrivileges = this.authorization!.checkPrivilegesWithRequest(this.request);
-    const privilege = privilegeFactory(this.authorization!);
-
-    const { username, privileges } = await checkPrivileges.atSpaces(spaceIds, {
-      kibana: privilege,
-    });
-    const authorized = privileges.kibana
-      .filter((x) => x.authorized)
-      .map((x) => x.resource as string);
-
-    this.debugLogger(
-      `SpacesClient.getAll(), authorized for ${
-        authorized.length
-      } spaces, derived from ES privilege check: ${JSON.stringify(privileges)}`
-    );
-
-    return { username, authorized };
-  }
-
   private async ensureAuthorizedGlobally(action: string, method: string, forbiddenMessage: string) {
     const checkPrivileges = this.authorization!.checkPrivilegesWithRequest(this.request);
     const { username, hasAllRequested } = await checkPrivileges.globally({ kibana: action });

```

Reply:
Done in 836ef49!